### DISCUSSION TITLE
[Feature] export to sidecar files

### DISCUSSION THREAD

**Discussion by gardiol** [immich-app/immich#10281] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

Long time (lifetime) support is critical to store a lifetime of photos. I have assets back to 25 years ago and they all survived today only because i am storing them sorted by folders and with metadata on sidecar files. I went trough many tools in that timespan and while Immich is awesome, nobody can ensure in 20+ years it will still be around, maintained and supported. While i whish it will be, i don't feel confident in wasting my time and effort in storing metadata (tags, names, faces, etc) "only" in a database. 

Even if is true that a database can always be dumped, it's also true that would be a massive pain and not so easy for the average photo amateur. 

It would be an anazing addition if Immich could do an export do common sidecar (xml) format of each picture metadata alongside the picture itself (or in a parallel tree, i don't really care as far as the tree structure is identical). This could be triggered by a task or done automatically.

### Platform

- [ ] Server
- [ ] Web
- [ ] Mobile

**Comment by bo0tzz** [comment]
Immich already writes metadata edits (datetime, location) to a sidecar. There are existing feature requests for writing other fields.

**Reply by athulpa** [reply]
I had to find out more by testing things out. Here's what I learned, as of v1.124 of immich:

 Immich will write the following metadata fields to a sidecar file **_if and when you make any changes_**:
- Description
- Date and time
- GPS Location
- Tags
(these are the editable fields in the info pane when viewing an image)

If any metadata edits are made via API calls (for example: when using a tool like immich-go), this behavior is the same as if a user has edited it manually.

If a sidecar file is already present, whether created by immich or originally by the user, immich will edit that file.
And if no sidecar file exists when you edit metadata, then immich creates this file. Naming will be _abc.jpg.xmp_ for an image _abc.jpg_.

**Reply by LuucE** [reply]
> Immich already writes metadata edits (datetime, location) to a sidecar. There are existing feature requests for writing other fields.

Yes, Immich does write metadata to sidecars, but the request is the feature to **export** the sidecars with the photos. 

Regards Luuc.

**Reply by 5TDg9** [reply]
Yes, just realized this is a lacking feature!  A lot of immich users, use digikam, too!
The information of Tags, desctiprion, ratings, etc should be easily exported.
As this information is in the XMP file already, this is not a big thing to do.

**Reply by Dylan10126** [reply]
> Yes, just realized this is a lacking feature! A lot of immich users, use digikam, too! The information of Tags, desctiprion, ratings, etc should be easily exported. As this information is in the XMP file already, this is not a big thing to do.

But depending on the situation, an XMP file may not even be created at all. If I perform a mobile upload, it will only extract the embedded metadata and that's it. Id argue that the average user will only apply people via facial frames; none of which is actually written to or will generate an XMP file either (unlike lightroom which adds both a people tag, and saves facial frames to metadata). 

Should someone suffer a database failure (no backup) or simply wish to change platforms, they would need to start from scratch if all they care about is people tagging. It's why I generally use external libraries only and run my photos through lightroom beforehand to get the improved metadata.

Its a shame since I find the Immich AI to be more accurate

**Reply by 5TDg9** [reply]
Thanks for the tip @Dylan10126 ! 
It's not just about face tags. Ratings, descriptions, and other EXIF ​​data should also be easy to export.

Closely related to this topic is the ability to search using tags and operators. PhotoPrism seems to have reached a somewhat more advanced stage of development here, according to [#874](https://github.com/photoprism/photoprism/discussions/874)

It's certainly obvious to everyone that tags are generally much better suited to sorting large amounts of images or files than traditional albums or folders.

For example, I use tags like #todo for images that I want to edit later, or tags like #milestone or #person, #landscape, #object, etc.

If you're wondering what's more important from a user perspective, after a moment's consideration, you'll come to the following conclusion: EXIF ​​interoperability, including tags and star ratings, is more important than searching for them, so that a locked-in issue doesn't arise at some point. But perhaps this issue isn't a top priority for developers, or they'd rather reserve this option for possible future monetization opportunities.

**Comment by Jrod696** [comment]
This would be a nice option to dump the db to XMP files, through a very dumb code mistype of my own I nuked about 80% of  my XMP files from a library that has several hundred thousand images in it and now they are just gone, right now its not a big deal because they are in the Immich DB but I have no way to recreate all the files I manually added dates and locations to, its like the EXIF data is being held hostage, my plan was to fill in all the info for the images and then run exiftool to write the xmp data to the files and then I would be set but that is where my code mishap went pear shaped and wiped out/corrupted my XMP files. 

To be clear although it would be nice to have more complicated XMP data, I am just looking for what it already creates now which is created date, modified date, and location, anything else like faces or tags would be a bonus but not crucial, just simple date and location takes care of any import needed should I need to do a recovery

**Reply by peemot** [reply]
Hi! Take a look at the script, it may help you: [immich-export](https://github.com/peemot/immich-export)
For now it does filter out only photos with face data. But the same approach could be used to export information for every photo.

**Reply by skatsubo** [reply]
@Jrod696 Back in December I wrote a bash+sql+exiftool script to export metadata to XMP sidecars: 

[immich-metadata-exporter](https://github.com/skatsubo/immich-metadata-exporter)

It's meant to be run on the Immich server: `bash export.sh --preview`. In preview (dry-run) mode it generates and writes all XMP files in the current directory, recreating the file system layout of original assets in a subdirectory.

(I've updated it recently to incorporate Immich DB schema changes - as the tool fetches metadata from the database directly, without API).

---

@peemot It's nice to see that `immich-export` supports writing virtually all metadata, not only faces, great!
